lib/lhm/sql_retry.rb

```diff
@@ -15,22 +15,124 @@ module Lhm
   # https://github.com/kamui/retriable. Additionally, a "log_prefix" option,
   # which is unique to SqlRetry can be used to prefix log output.
   class SqlRetry
-    def initialize(connection, options = {})
+
+    RECONNECT_SUCCESSFUL_MESSAGE = "LHM successfully reconnected to initial host:"
+    ABNORMAL_EXECUTION_TIME_THRESHOLD = 5
+
+    def initialize(connection, options = {}, with_consistent_host = true, retry_for_host_options = {})
       @connection = connection
       @log_prefix = options.delete(:log_prefix)
       @retry_config = default_retry_config.dup.merge!(options)
+      if (@with_consistent_host = with_consistent_host)
+        @host_retry_config = default_host_retry_config.dup.merge!(retry_for_host_options)
+        @initial_host = hostname
+        @initial_server_id = server_id
+      end
     end
 
-    def with_retries
+    # Complete explanation of algorithm: https://github.com/Shopify/db-engineering/issues/98#issuecomment-934948590
+    def with_retries(override_prefix=nil)
+      # Overrides log prefix if necessary or passed from parent to child instance (ex: Chunker -> ChunkInsert)
+      old_prefix =  @log_prefix
+      @log_prefix = override_prefix unless override_prefix.nil?
+
       Retriable.retriable(retry_config) do
-        yield(@connection)
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+        # The time check will log a warning if an action takes more than 5s (there's a possibility a
+        # failover might've happened after the hostname check above)
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        # Since Lhm::Error < StandardError, it is important to separate the flows, otherwise every
+        # error would trigger a reconnect. In this case, some instances of Lhm::Error are raised in a controlled manner
+        # to trigger retries from ` Retriable.retriable`
```

Comment:
I had to read this comment a few times and still don't really get it. If some "some instances" of Lhm::Error trigger a controlled retry, why not raise something like `TryReconnect` instead of `LhmError`?